In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import sklearn

In [4]:
df = pd.read_csv('kc_house_data.csv')
df.head()

id             date     price  ...     long  sqft_living15  sqft_lot15
0  7129300520  20141013T000000  221900.0  ... -122.257           1340        5650
1  6414100192  20141209T000000  538000.0  ... -122.319           1690        7639
2  5631500400  20150225T000000  180000.0  ... -122.233           2720        8062
3  2487200875  20141209T000000  604000.0  ... -122.393           1360        5000
4  1954400510  20150218T000000  510000.0  ... -122.045           1800        7503

[5 rows x 21 columns]

In [6]:
X = pd.DataFrame(df[['bedrooms','bathrooms','sqft_living','sqft_lot','floors','waterfront','view','sqft_above','sqft_basement','sqft_living15','sqft_lot15','condition','grade']])
Y = pd.Series(df['price'])
X.head()

bedrooms  bathrooms  sqft_living  ...  sqft_lot15  condition  grade
0         3       1.00         1180  ...        5650          3      7
1         3       2.25         2570  ...        7639          3      7
2         2       1.00          770  ...        8062          3      6
3         4       3.00         1960  ...        5000          5      7
4         3       2.00         1680  ...        7503          3      8

[5 rows x 13 columns]

##Encoding Ordinal Categorical Data like Condition and Grade

In [ ]:
from sklearn.preprocessing import OrdinalEncoder
enc = OrdinalEncoder()
enc.fit(X)

##LINEAR REGRESSION

In [10]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size = 0.2,random_state=5)

In [11]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train,Y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [12]:
y_pred = regressor.predict(X_test)
y_pred

array([588289.87421617, 201225.313863  , 406878.92668003, ...,
       554659.49038423, 703364.56484755, 549131.10886254])

In [13]:
from sklearn.metrics import r2_score
r2_score(Y_test,y_pred)

0.6111480391176247

##POLYNOMIAL REGRESSION


In [14]:
from sklearn.preprocessing import PolynomialFeatures
poly = PolynomialFeatures(degree=2)
x_poly = poly.fit_transform(X)

In [15]:
X_train,X_test,Y_train,Y_test = train_test_split(x_poly,Y,test_size = 0.2,random_state=5)
regressor = LinearRegression()
regressor.fit(X_train,Y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [16]:
y_pred = regressor.predict(X_test)
y_pred

array([538265.98740036, 326170.56650664, 337961.43408801, ...,
       538440.68799056, 732066.60160907, 565879.92720046])

In [17]:
from sklearn.metrics import r2_score
r2_score(Y_test,y_pred)

0.6886751005111312

##DECISION TREE REGRESSION

In [18]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size = 0.2,random_state=5)

In [19]:
from sklearn.tree import DecisionTreeRegressor
regressor_3 = DecisionTreeRegressor(random_state = 5)
regressor_3.fit(X_train,Y_train)

DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=None,
                      max_features=None, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, presort='deprecated',
                      random_state=5, splitter='best')

In [20]:
y_pred = regressor_3.predict(X_test)
y_pred

array([424950., 227000., 386500., ..., 415000., 599000., 572500.])

In [21]:
from sklearn.metrics import r2_score
r2_score(Y_test,y_pred)

0.45374060169238906

##RANDOM FOREST REGRESSION

In [22]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size = 0.2,random_state=5)

In [23]:
from sklearn.ensemble import RandomForestRegressor
regressor_4 = RandomForestRegressor(n_estimators = 200)
regressor_4.fit(X_train,Y_train)

RandomForestRegressor(bootstrap=True, ccp_alpha=0.0, criterion='mse',
                      max_depth=None, max_features='auto', max_leaf_nodes=None,
                      max_samples=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      n_estimators=200, n_jobs=None, oob_score=False,
                      random_state=None, verbose=0, warm_start=False)

In [24]:
y_pred = regressor_4.predict(X_test)
y_pred

array([521116.75 , 255738.025, 335449.49 , ..., 437162.85 , 607677.46 ,
       503175.415])

In [25]:
from sklearn.metrics import r2_score
r2_score(Y_test,y_pred)

0.7271851530528699

##SUPPORT VECTOR REGRESSION

In [27]:
Y = df.iloc[:, 2].values
Y = Y.reshape(len(Y),1)

In [28]:
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
sc_y = StandardScaler()
X = sc_X.fit_transform(X)
Y = sc_y.fit_transform(Y)

In [29]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size = 0.2,random_state=5)

In [30]:
from sklearn.svm import SVR
regressor_5 = SVR(kernel = 'rbf')
regressor_5.fit(X_train, Y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVR(C=1.0, cache_size=200, coef0=0.0, degree=3, epsilon=0.1, gamma='scale',
    kernel='rbf', max_iter=-1, shrinking=True, tol=0.001, verbose=False)

In [31]:
y_pred = regressor_5.predict(X_test)
y_pred

array([-0.31301582, -0.70089302, -0.6170411 , ..., -0.11875143,
        0.38503874,  0.02563411])

In [32]:
from sklearn.metrics import r2_score
r2_score(Y_test,y_pred)

0.6815932623010323

####Looking at all R-squared scores,Random Forest Regression has the highest value, that is, 0.7271851530528699. Hence Random Forest Regression is the best model in this case

####Improvements can be made by doing more feature engineering on the dataset.